docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+  "content-type": "application/json",
+  “error”: {
+    "message": "an error occurred, blah, blah",
+    "type": "Exception",
+    "trace": "stack trace"
+  },
+  “extensions”: {
+    “protocol”: {
+       “headers”: …
+       Etc, same as we do now for JSON format. 
+    }
+  }
+}
+```
+
+Error map:
+* message - required
+* type - type of error, eg IOError - optional
+* trace - stack trace - optional
+
+### Triggers
+
+Triggers bind events to functions along with configuration.
+
+Eg: HTTP event
+
+```json
+{
+    "source": "http://myapp.com/x"
+}
+```
+
+Trigger will look up the source in the triggers db then update the event with the Function information plus the app and trigger configuration.
+
+```json
+{
+    "extensions": {
+        "trigger": {
+            "function_url": "func url",
+            "config": {
+                MAP OF CONFIGS FROM APP AND TRIGGER
+            }
+        }
+    }
+}
+```
+
+It will then pass this message along to the Fn Router/LB for execution.
+
+### Apps/Applications
+
+Applications are a grouping of triggers to make up a complete application. For example, an application might have the following triggers:
+
+* http: /x -> trigger maps /x -> function f1 with config c1
+* http: /y …
+* schedule: run function f2 every day at 8am
+* queue: listen to topic t1 filtering on ff1, for every event run function f3
+
+All of these things make up the application. Applications also define 
+additional annotations and configuration to be added to each trigger.
+
+## Components
+
+The following components make up the Fn platform. 
+
+* All components can run separately
+* They all have their own main
+* One main main to tie them all into one bin for a nice developer experience.
+
+### Runner
+
+The Runner does nothing more than take an event message, execute the function and return results.
+
+* Main endpoint `/run` with CloudEvent as input.
+* When called, runs a function and returns the results, as a CloudEvent.
+* If busy, returns 503.
+* Event input contains everything it needs to run, including config/secrets.
+  * TODO: Need to be sure we don’t log any secrets
+* Logs can be retrieved separately via `/log` endpoint so main response can be returned faster.
+
+### Function Registry
+
+A service that gives functions an addressable URL, eg:
+
+```
+https://functions.somewhere.com/NAMESPACE/FUNCNAME:VERSION-TAG
+```
+
+* Has metadata required to run a function provided by the creator of the function
+* All functions registered here must accept and return CloudEvent messages
+* This leads to reusable/shareable functions with no dependencies on other components
+* Could potentially run as a separate service like Docker Hub
+
+Q: Seems like a lot to just be a small layer on top of Docker Hub?
+A: It may seem this way now for just a few fields, but down the road, there could many additional things
+  like forcing functions to pass certain tests (eg: I/O formats), defining expected input parameters (can reject 
```

Comment:
FunctionsHub is not just a docker registry. If we'd talk about public hub - it has to be similar to Docker EE hub with trusted function.

I see a FunctionsHub as a combination of Github and DockerHub. You need to see function's source code and the way it's packaged (would that be just a func file or Dockerfile).

Reply:
Could even make it so that the metadata is in a file in the git repo and its only accepted to the Functions hub if it is in the git repo.. Then we can use any git repo link + docker image link. Then it can be deployed on prem and use private repos / images or on public cloud. Just my 2 cents.

Reply:
This could be private or public and source code doesn't need to be available.

Reply:
Right so I meant that git repo could be optional but to not locking on requiring docker hub specifically. Just the meta data and the docker image. But having a git repo that is linked to it as an option would be very useful :)

Reply:
imo it was kind of a non-goal to design a functions hub here. we just need to split functions out from the way we have it now, and have triggers set up to be able to address functions by name. we can design function hub down the road, but is it nice contextually to have that in mind here and it shows how powerful this abstraction could be.